### PR TITLE
Merge DSM context extraction and reduce its public API

### PR DESCRIPTION
**Pull request by PerfectSlayer**
# What Does This Do

This PR merges the DSM context extraction with the default span extraction.
It also reduces the DSM public API to the minimum needed and remove public API related to the Span.Context mutation.

# Motivation

Merging context extraction limits the creation and mutation of extracted context (using a one-pass creation).
It should help with the future context API and decoupling DSM context from Span and its context.

# Additional Notes

### CODE REVIEW

**Review by bantonsson — APPROVED** [review]
Nice. We should probably get a set of eyes from someone that worked on DSM as well.

**Review by nayeem-kamal — APPROVED** [review]

**Review by hokitam — COMMENTED** [review]
Thank you for doing this refactoring! I have some small concerns about the separation of context extraction from the setDataStreamsCheckpoint call which may make it easier to introduce some null pointer exceptions down the line if we don't add some null checks now. LMK what you think!

File: dd-trace-core/src/main/java/datadog/trace/core/datastreams/DefaultPathwayContext.java [hunk 1]
```diff
@@ -279,6 +279,10 @@ public static <C> DefaultPathwayContext extract(
       AgentPropagation.ContextVisitor<C> getter,
       TimeSource timeSource,
       WellKnownTags wellKnownTags) {
+    if (getter instanceof AgentPropagation.BinaryContextVisitor) {
+      return extractBinary(
```
[the remark this answers is not in the document]
Comment (PerfectSlayer) [reply dd-trace-core/src/main/java/datadog/trace/core/datastreams/DefaultPathwayContext.java:283]: I did not plan to reduce API surface but I can based on your agreement.
So `extract()` and `extractBinary()` will no be package private (`extract()` is still used by `DefaultDatastreamMonitoring` and it feels better they have the same scope).
I also set all three decode methods to private (`decode`, `strDecode` and `base64Decode`).

File: dd-trace-core/src/main/java/datadog/trace/core/CoreTracer.java [hunk 2]
```diff
@@ -818,7 +823,7 @@ public <C> void inject(AgentSpan span, C carrier, Setter<C> setter, TracePropaga
   public <C> void injectBinaryPathwayContext(
       AgentSpan span, C carrier, BinarySetter<C> setter, LinkedHashMap<String, String> sortedTags) {
     PathwayContext pathwayContext = span.context().getPathwayContext();
-    pathwayContext.setCheckpoint(sortedTags, dataStreamsMonitoring);
+    pathwayContext.setCheckpoint(sortedTags, dataStreamsMonitoring::add);
```
[the remark this answers is not in the document]
Comment (PerfectSlayer) [reply dd-trace-core/src/main/java/datadog/trace/core/CoreTracer.java:826]: Good idea. I checked and protected all calls to `getPathwayContext()` with `null` check.
I am not in favor about returning `null` values and null checks but the API was design that way and I would better not change to many things in this refactoring.

**Review by hokitam — APPROVED** [review]
Thank you for the updates!

**Comment by hokitam** [comment]
one more question: Why did you separate out `AgentDataStreamsMonitoring` interface from the `DataStreamsMonitoring` one? Is it mainly due to which functionalities were getting used in which packages?

**Pull request merged**